test/gtest/ucp/test_ucp_wireup.cc

```diff
@@ -809,3 +817,93 @@ UCS_TEST_P(test_ucp_wireup_errh_peer, msg_before_ep_create) {
+    enum_test_params(const ucp_params_t& ctx_params, const std::string& name,
+                     const std::string& test_case_name, const std::string& tls)
+    {
+        return enum_test_params_features(ctx_params, name, test_case_name, tls,
+                                                                                  UCP_FEATURE_TAG | UCP_FEATURE_RMA |
+                                         UCP_FEATURE_STREAM, 1);
+    }
+
+    void init() {
+        /* do nothing */
+    }
+
+    void cleanup() {
+        /* do nothing */
+    }
+
+    bool test_est_num_eps_fallback(size_t est_num_eps,
+                                   size_t &min_max_num_eps) {
+        bool res = true;
+        min_max_num_eps = UCS_ULUNITS_INF;
+
+        UCS_TEST_MESSAGE << "Testing " << est_num_eps << " number of EPs";
+        modify_config("NUM_EPS", ucs::to_string(est_num_eps).c_str());
+        test_ucp_wireup::init();
+
+        sender().connect(&receiver(), get_ep_params());
+        if (!is_loopback()) {
+            receiver().connect(&sender(), get_ep_params());
+        }
+        send_recv(sender().ep(), receiver().worker(), receiver().ep(), 1, 1);
+        flush_worker(sender());
+
+        for (ucp_lane_index_t lane = 0;
+             lane < ucp_ep_num_lanes(sender().ep()); lane++) {
+            uct_ep_h uct_ep = sender().ep()->uct_eps[lane];
+            if (uct_ep == NULL) {
+                continue;
+            }
+
+            uct_iface_attr_t iface_attr;
+            ucs_status_t status = uct_iface_query(uct_ep->iface, &iface_attr);
+            ASSERT_UCS_OK(status);
+
+            if (iface_attr.max_num_eps < est_num_eps) {
+                res = false;
+                goto out;
+            }
+
+            if (iface_attr.max_num_eps < min_max_num_eps) {
+                min_max_num_eps = iface_attr.max_num_eps;
+            }
+        }
+
+    out:
+        test_ucp_wireup::cleanup();
+        return res;
+    }
+};
+
+UCS_TEST_P(test_ucp_wireup_fallback, est_num_eps_fallback) {
+    size_t test_min_max_eps, min_max_eps;
+
+    test_est_num_eps_fallback(1, test_min_max_eps);
+
+    while (test_min_max_eps != UCS_ULUNITS_INF) {
+        if (test_min_max_eps > 1) {
+            EXPECT_TRUE(test_est_num_eps_fallback(test_min_max_eps - 1,
+                                                  min_max_eps));
+        }
+
+        EXPECT_TRUE(test_est_num_eps_fallback(test_min_max_eps,
+                                              min_max_eps));
+
+        EXPECT_TRUE(test_est_num_eps_fallback(test_min_max_eps + 1,
+                                              min_max_eps));
+        test_min_max_eps = min_max_eps;
+    }
+}
+
+/* Test fallback from RC to UD, since RC isn't scalable enough
+ * as its iface max_num_eps attribute = 256 by default */
+UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_wireup_fallback,
+                              rc_ud, "rc_x,rc_v,ud_x,ud_v")
+/* Test two scalable enough transports */
```

Comment:
maybe also test single unscalable transport to make sure all lanes selected properly rather than ep creation fails?

Reply:
> maybe also test single unscalable transport to make sure all lanes selected properly rather than ep creation fails?

yes, this is a good idea, I'm looking for a good way on how we could do that